### ISSUE TITLE
Flattening loss on the validation set.

### ISSUE THREAD

**Issue by RadoslavMYankov** [theislab/FlatVI#2]
Thank you very much for your amazing work!

While experimenting with FlatVI I noticed that the flattening loss explodes and is multiple orders of magnitude higher compared to the training loss. Have you also observed this behaviour and do you prefer other metrics to evaluate the flatness of the latent space?

**Comment by allepalma** [comment]
Hi @RadoslavMYankov, thanks for your interest. What scale did you assign to it? Does the loss decrease or increase steadily? 

You could start downscaling it a bit; it helps. Let me know if it works for you!

**Comment by RadoslavMYankov** [comment]
Hi @allepalma, thanks for your reply! I realized that I have phrased my question horribly... Apologies for that. I have run multiple experiments with FlatVI and varying scales of the fl_weigh (0.1, 1.0, and 10.0) both as a constant value and with annealing strategies (incremental or decremental). The discrepancy that I was referring to in the original question is between the training and the validation values of the flattening loss. In all cases training converges smoothly. The problem is that while the training loss is decreasing the validation loss is increasing (again I am only referring to the geometric loss here). 

<img width="1400" height="578" alt="Image" src="https://github.com/user-attachments/assets/5196c3ae-5a6d-4e68-8bc0-8ec3cca50c82" />

<img width="1418" height="601" alt="Image" src="https://github.com/user-attachments/assets/f45366b4-ee58-431e-bcd6-6d126910b1ca" />

I ran some sanity checks and it seems like the higher the scale the better the generalization, regardless in some cases the flattening loss remains exceptionally high in the range of 10^9-12 and having weights higher than 10.0 seems prohibitive for the reconstruction loss. 

I also ran some tests (on eb as well as other datasets) where I simply logged the loss without using it for gradient updates and in this case the flattening loss explodes to 10^25. I guess my initial intuition is incorrect and FlatVI does generalize even for the validation dataset. However, the extent to which it is able to stretch out the latent space in areas without support points is limited.

I am still concrened about the validation loss increasing as training progresses and the training loss converges nicely. My intuition is to compare the latent space with a crumbled sheet of paper where flattening/smoothening out a segment would inadvertently crumble another one. The other problem is that even after the sanity checks I observe a discrepancy of at least 4 orders of magnitude between the training and validation loss (best case and only a single experiment with scale 10) and usually much higer (at least 6-7 degrees of magnitude).

For this reason I wanted to ask if you have observed similar behaviour and if this is expected or should I look for bugs in my code. I have compared my loaders and preprocessing with your implementation and obtained the same results but I may still be missing something.

I find your work really intriguing and I also understand that stretching out the latent space which is very far from euclidean is challenging and that the paper never claims to do it completely. I just want to confirm that my initial experiments are not too far from your results and thus be able to carefully consider my next steps.

Thank you and best regards,
Rado

**Comment by allepalma** [comment]
Hi @RadoslavMYankov. Sorry, I got caught up in a deadline! I think your observation is correct with respect to magnitude. I had noticed a similar behaviour, especially in datasets with a limited number of data points. For example, I had initially tried the model on an infection dataset with a few datapoints and noticed this effect was more pronounced as the manifold is undersampled. What dataset did you try this on? Do you enable learning a diagonal scaling constant?

**Comment by RadoslavMYankov** [comment]
Hi @allepalma, thank you for your reply! I hope your project will be successful. I am trying to integrate some spatial information in the autoencoder and noticed some marginal improvements on the downstream task. However the scale discrepancy is even larger when working with my extension. The dataset that I am looking into for these experiments is the Axolotl brain development dataset (Wei et al). C is enabled in my configs. However, I see similar behaviour on the EB dataset from the paper.

**Comment by allepalma** [comment]
Dear @RadoslavMYankov, thanks a lot for your continued interest! The magnitude effect is something I'd expect. As you suggest, this matrix-comparison-based loss is very sensitive, so it will attain high values very easily when you leave the support. This effect is amplified by high dimensions, but in my experience still produces reliable results in practice when you inspect the latents. :)   

However, in my case, it has more of a double trend. It increases at the beginning but then goes down a bit after a while. What latent size have you used? I expect it to increase at the beginning as the model is overfitting the training data and building the latent manifold. But then it should indeed decrease a bit.

On another note, did you inspect the model on your downstream task?

**Comment by RadoslavMYankov** [comment]
@allepalma Thanks for the swift reply! Your suggestions make sense and perhaps I could run a few more experiments on the downstream task to get a better understanding.

So far I have stuck with the values reported in the paper - 2 layers width width 256 and 10 respectively. Finally, if you would be open for it, could you provide some insights on inspecting the latents after generating the embeddings, or during training on the downstream task?

Would you consider a small improvement a sufficient indication of performance in spite of the magnitude of the loss?

**Comment by allepalma** [comment]
Hi @RadoslavMYankov, depends a bit :) 

What task are you broadly tackling? Does Euclideanity in the latent space help in theory?

**Comment by RadoslavMYankov** [comment]
@allepalma Apologies for the late reply. I am also using flow matching for trajectory inference. I am still exploring different representations to find the most suitable approach and an euclidean latent space would fit neatly with OT CFM. Nevertheless, there might be alternative FM objective formulations that are better suited to tackling this issue in isolation as FlatVI only presents this as a potential use case but the main objective is to regularize the latent space.

**Comment by allepalma** [comment]
@RadoslavMYankov  This sounds great! Let me know if you end up using FlatVI's representations for your task and if they are useful. 

My understanding is that the magnitude difference is normal, but the increase in loss worries me a bit, as I did not experience it.  I think further research could definitely be invested in experimenting with methods for flattening the manifold a bit more uniformly. Not sure how easy it is in high-dimensions, as even in 10-d the single-cell manifold is essentially undersampled. I think larger datasets and as low a low-d scenario as possible are the best approaches. 

I will close the issue now, but I will look more into it. If you come up with enhancements, I'd be very happy to hear about them and potentially implement them! In the meantime, good luck with your work and thanks again for reaching out! Feel free to open new issues if new problems arise!
